### PR TITLE
bpo-38091: Import deadlock detection causes deadlock

### PR DESCRIPTION
**Pull request by arigo**
https://bugs.python.org/issue38091

Automerge-Triggered-By: @brettcannon

### CODE REVIEW

**Comment by pitrou** [comment]
The fix here is more of a hack. A proper fix could be something like the below. At least it fixes the reproducer posted by @rlamy . Is there a real-world reproducer you can test it on?

```diff
diff --git a/Lib/importlib/_bootstrap.py b/Lib/importlib/_bootstrap.py
index 8de0e9ee79..7b2bfefdec 100644
--- a/Lib/importlib/_bootstrap.py
+++ b/Lib/importlib/_bootstrap.py
@@ -71,9 +71,20 @@ class _ModuleLock:
             lock = _blocking_on.get(tid)
             if lock is None:
                 return False
-            tid = lock.owner
-            if tid == me:
+            owner_tid = lock.owner
+            if owner_tid == tid:
+                # Lock owner is temporarily blocking on a lock it
+                # already owns.  This means it's not actually blocking.
+                # (this can happen at the beginning of acquire() below
+                #  -- see bpo 38091)
+                return False
+            if owner_tid == me:
+                # Lock owner is blocking on a lock we own, but we're blocking
+                # on the lock it owns => deadlock.
                 return True
+            # Move along the chain and find out whether the owner is blocking
+            # on another lock.
+            tid = owner_tid
 
     def acquire(self):
         """
```

**Comment by arigo** [comment]
Maybe a safer way to think about it would be to ask ourselves when has_deadlock() is really supposed to return True.  I think it means to return True if the chain ``_blocking_on[_blocking_on[...[tid]...]]`` eventually reaches the value ``me``, and False in all other cases.  You are fixing one such case by detecting when the chain enters a fixpoint, i.e. a loop of length 1.  I'm not convinced that longer loops are impossible; maybe we should also detect them and return False in these cases.  Something like that (untested code):

```
     def has_deadlock(self):
         # Deadlock avoidance for concurrent circular imports.
         me = _thread.get_ident()
         tid = self.owner
+        seen = set()
         while True:
             lock = _blocking_on.get(tid)
             if lock is None:
                 return False
             tid = lock.owner
             if tid == me:
                 return True
+            if tid in seen:
+                # bpo 38091: the chain of tid's we encounter here
+                # eventually leads to a fixpoint or a cycle, but
+                # does not reach 'me'.  This means we would not
+                # actually deadlock.  This can happen if other
+                # threads are at the beginning of acquire() below.
+                return False    
+            seen.add(tid)
```

**Comment by pitrou** [comment]
@arigo Yes, that fix would probably work. Perhaps @rlamy can check :-)

**Comment by rlamy** [comment]
I think that this should fix the issue. It caused some intermittent test failures in PyPy, so I guess we should just merge the change there and monitor the CI results.

**Comment by pitrou** [comment]
@rlamy Did you test this fix on PyPy?

**Review by brettcannon — CHANGES_REQUESTED** [review]
Marking as rejected until we here back from PyPy that the proposed fix in the PR discussion worked for them.

**Comment by arigo** [comment]
I have made the requested changes; please review again

**Comment by arigo** [comment]
I have updated the PR to include the test from https://bugs.python.org/issue38091 , which fails (deadlocks) without the fix and passes with the fix.

**Review by brettcannon — APPROVED** [review]

**Review by brettcannon — CHANGES_REQUESTED** [review]
`importlib.h` wasn't regenerated and included in this PR. I believe that can be done with `make regen-all` (or something like that).

**Comment by brettcannon** [comment]
The change looks good, but `importlib.h` updates accidentally got left out. There is also no news entry (you can use https://blurb-it.herokuapp.com/ to do that).

**Comment by brettcannon** [comment]
BTW I'm assuming the test failures are due to `importlib.h` not getting updated in the PR.

**Comment by arigo** [comment]
I have made the requested changes; please review again

File: Lib/test/test_import/__init__.py [hunk 1]
```diff
@@ -436,16 +436,24 @@ def test_issue31492(self):
                 os.does_not_exist
 
     def test_concurrency(self):
+        # bpo 38091: this is a hack to slow down the code that calls
+        # has_deadlock(); the logic was itself sometimes deadlocking
```
Comment (brettcannon) [inline Lib/test/test_import/__init__.py]:
```suggestion
        # has_deadlock(); the logic was itself sometimes deadlocking.
```

**Review by brettcannon — APPROVED** [review]

File: Misc/NEWS.d/next/Core and Builtins/2020-02-14-23-10-07.bpo-38091.pwR0K7.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+The import logic has got code to detect deadlock. This code was faulty. It occasionally caused deadlocks to occur.
```
Comment (brettcannon) [inline Misc/NEWS.d/next/Core and Builtins/2020-02-14-23-10-07.bpo-38091.pwR0K7.rst]:
```suggestion
Tweak import deadlock detection code to not deadlock itself.
```

**Comment by brettcannon** [comment]
@arigo any interest in manually backporting this to 3.8 and 3.7? (Chances are it's `importlib.h`.)

**Comment by arigo** [comment]
My git-fu is very limited (in general and on github).  Do you expect two other pull requests for the 3.7 and 3.8 branches?

**Comment by brettcannon** [comment]
@arigo yep, but don't worry about it too much. I can try to get around to it eventually myself.

**Pull request merged**